### PR TITLE
VLN Task & Dataset with shortest path example and tests.

### PR DESCRIPTION
**Pull request by koshyanand**
## Motivation and Context

- [x] Add VLN Task 

- [x] Add reading of Room-to-Room Dataset

- [x] Add VLN Instruction Sensor

- [x] Added shortest path follower 

- [x] Added benchmark for Random, ForwardOnly, RandomForward & GoalFollower agent

- [x] Represent instructions as list of tokens

- [x] Added vocabulary API to R2R dataset

## How Has This Been Tested

- [x] Config test

- [x] Dataset size and serialization test

- [x] Instruction Sensor test & test to see if the sensor instruction matches the current instruction

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: habitat/tasks/vln/vln.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        return self._get_observation(**kwargs)
+
+
+@registry.register_task(name="VLN-v0")
+class VLNTask(NavigationTask):
+    _sensor_suite: SensorSuite
+
+    def __init__(self, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self._sensor_suite = SensorSuite([InstructionSensor()])
```
Comment (erikwijmans) [inline habitat/tasks/vln/vln.py]: This shouldn't be needed.

File: habitat/tasks/utils.py [hunk 2]
```diff
@@ -64,3 +64,34 @@ def cartesian_to_polar(x, y):
     rho = np.sqrt(x ** 2 + y ** 2)
     phi = np.arctan2(y, x)
     return rho, phi
+
+
+def quat_from_two_vectors(v0: np.ndarray, v1: np.ndarray) -> np.quaternion:
```
Comment (erikwijmans) [inline habitat/tasks/utils.py]: Is this needed anywhere?  Or can it be kept just within habitat-sim?

File: configs/tasks/vln_r2r.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    TYPE: HabitatSimRGBSensor
+
+  DEPTH_SENSOR:
+    WIDTH: 256
+    HEIGHT: 256
+TASK:
+  TYPE: VLN-v0
+  SUCCESS_DISTANCE: 3.0
+  #Added pointgoal sensor for testing shortest path
+  SENSORS: ['POINTGOAL_WITH_GPS_COMPASS_SENSOR', 'INSTRUCTION_SENSOR']
```
Comment (erikwijmans) [inline configs/tasks/vln_r2r.yaml]: The PointGoal sensor should be added just in that test.

File: configs/datasets/vln/mp3d_r2r.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+DATASET:
+  TYPE: R2RVLN-v1
+  SPLIT: train
+  DATA_PATH: data/R2R/habitat_R2R_{split}.json
```
Comment (mathfac) [inline configs/datasets/vln/mp3d_r2r.yaml]: To follow naming dataset files naming convention I would advice next naming:
```suggestion
  DATA_PATH:   data/datasets/r2r/{split}/{split}.json.gz
```

File: configs/tasks/vln_r2r.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+ENVIRONMENT:
+  MAX_EPISODE_STEPS: 500
+SIMULATOR:
+  FORWARD_STEP_SIZE: 0.25
+  TURN_ANGLE: 15
+  HABITAT_SIM_V0:
+    GPU_DEVICE_ID: 0
+  RGB_SENSOR:
+    WIDTH: 512
```
Comment (mathfac) [inline configs/tasks/vln_r2r.yaml]: Is that expected that RGB resolution is different from Depth resolution?

File: configs/datasets/vln/mp3d_r2r.yaml [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+  SCENES_DIR: data/mp3d/scenes/
```
Comment (mathfac) [inline configs/datasets/vln/mp3d_r2r.yaml]: Default path is expected to be:
```suggestion
  SCENES_DIR: "data/scene_datasets/"
```

File: configs/tasks/vln_r2r.yaml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  SENSORS: ['INSTRUCTION_SENSOR']
+  MEASUREMENTS: ['SPL']
+  SPL:
+    TYPE: SPL
+    SUCCESS_DISTANCE: 2.0
+DATASET:
+  TYPE: R2RVLN-v1
+  SPLIT: train
+  DATA_PATH: data/datasets/R2R/hb_R2R_{split}.json
+  SCENES_DIR: data/scene_datasets/mp3d/
```
Comment (mathfac) [inline configs/tasks/vln_r2r.yaml]: Same comment regarding dataset paths.

File: configs/tasks/vln_r2r.yaml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    HEIGHT: 512
+    HFOV: 45
+    TYPE: HabitatSimRGBSensor
+  DEPTH_SENSOR:
+    WIDTH: 256
+    HEIGHT: 256
+TASK:
+  TYPE: VLN-v0
+  SUCCESS_DISTANCE: 2.0
+  SENSORS: ['INSTRUCTION_SENSOR']
```
Comment (mathfac) [inline configs/tasks/vln_r2r.yaml]: Worth to add explicitly supported actions:
` POSSIBLE_ACTIONS: ['MOVE_FORWARD', 'TURN_LEFT', 'TURN_RIGHT', 'STOP']`

File: configs/tasks/vln_r2r.yaml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    WIDTH: 512
+    HEIGHT: 512
+    HFOV: 45
+    TYPE: HabitatSimRGBSensor
+  DEPTH_SENSOR:
+    WIDTH: 256
+    HEIGHT: 256
+TASK:
+  TYPE: VLN-v0
+  SUCCESS_DISTANCE: 2.0
```
Comment (mathfac) [inline configs/tasks/vln_r2r.yaml]: 2 meters are acceptable success distance? Can be pretty large threshold.

File: configs/tasks/vln_r2r.yaml [hunk 10]
```diff
@@ -0,0 +1,28 @@
+ENVIRONMENT:
+  MAX_EPISODE_STEPS: 500
+SIMULATOR:
```
Comment (mathfac) [inline configs/tasks/vln_r2r.yaml]: Missed sensors enabling:
```
  AGENT_0:
    SENSORS: ['RGB_SENSOR', 'DEPTH_SENSOR']
```

File: configs/test/habitat_r2r_vln_test.yaml [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    DIMENSIONALITY: 2
+  GOAL_SENSOR_UUID: pointgoal_with_gps_compass
+  MEASUREMENTS: ['SPL']
+  SPL:
+    TYPE: SPL
+    SUCCESS_DISTANCE: 2.0
+DATASET:
+  TYPE: R2RVLN-v1
+  SPLIT: val_seen
+  DATA_PATH: data/datasets/R2R/hb_R2R_{split}.json
```
Comment (mathfac) [inline configs/test/habitat_r2r_vln_test.yaml]: Dataset files paths and naming.

File: configs/tasks/vln_r2r.yaml [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    WIDTH: 256
+    HEIGHT: 256
+    HFOV: 45
+    TYPE: HabitatSimRGBSensor
+  DEPTH_SENSOR:
+    WIDTH: 256
+    HEIGHT: 256
+TASK:
+  TYPE: VLN-v0
+  SUCCESS_DISTANCE: 3.0
```
Comment (jacobkrantz) [inline configs/tasks/vln_r2r.yaml]: The success distance is set to 3m here which is the distance defined in the original paper ([https://arxiv.org/abs/1711.07280](url)) and is used by most papers in this space.

File: configs/tasks/vln_r2r.yaml (hunk 8 above)
Comment (erikwijmans) [reply configs/tasks/vln_r2r.yaml]: Does stop need to go first to keep stop on 0?

Comment (mathfac) [reply configs/tasks/vln_r2r.yaml]: Correct, edited my comment.

File: test/test_r2r_vln.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    ), "JSON dataset encoding/decoding isn't consistent"
+
+
+def test_r2r_vln_dataset():
+    dataset_config = get_config(CFG_TEST).DATASET
+    if not r2r_vln_dataset.VLNDatasetV1.check_config_paths_exist(
+        dataset_config
+    ):
+        pytest.skip(
+            "Please download Matterport3D R2R dataset to " "data folder."
```
Comment (mathfac) [inline test/test_r2r_vln.py]:
```suggestion
            "Please download Matterport3D R2R dataset to data folder."
```

File: test/test_r2r_vln.py [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        )
+
+    dataset = r2r_vln_dataset.VLNDatasetV1(config=dataset_config)
```
Comment (mathfac) [inline test/test_r2r_vln.py]: Maybe use `make_dataset` factory call to create dataset.

File: test/test_r2r_vln.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        id_dataset=vln_config.DATASET.TYPE, config=vln_config.DATASET
+    )
+
+    env = SimpleRLEnv(config=vln_config, dataset=dataset)
+    env.episodes = dataset.episodes[:EPISODES_LIMIT]
+
+    follower = ShortestPathFollower(
+        env.habitat_env.sim, goal_radius=0.5, return_one_hot=False
+    )
+
```
Comment (mathfac) [inline test/test_r2r_vln.py]:
```suggestion
```

File: test/test_r2r_vln.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python3
```
Comment (mathfac) [inline test/test_r2r_vln.py]: Thank you for adding relevant tests.

File: habitat/tasks/vln/vln.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        goals: relevant goal object/room.
+    """
+    path: List[List[float]] = attr.ib(
+        default=None, validator=not_none_validator
+    )
+    instruction: InstructionData = attr.ib(
+        default=None, validator=not_none_validator
+    )
+    trajectory_id: int = attr.ib(default=None, validator=not_none_validator)
+    goals: List[NavigationGoal] = attr.ib(
```
Comment (mathfac) [inline habitat/tasks/vln/vln.py]: `goals` should be inherited from `NavigationEpisode` class.

File: habitat/tasks/vln/vln.py [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+        scene_id: id of scene inside the simulator.
+        start_position: numpy ndarray containing 3 entries for (x, y, z).
+        start_rotation: numpy ndarray with 4 entries for (x, y, z, w)
+            elements of unit quaternion (versor) representing agent 3D
+            orientation.
+        instruction: single instruction guide to goal.
+        trajectory_id: id of ground truth trajectory path.
+        goals: relevant goal object/room.
+    """
+    path: List[List[float]] = attr.ib(
```
Comment (mathfac) [inline habitat/tasks/vln/vln.py]: Can we save `path` as `shortest_path` in NavigationEpisode? Also doc string is missing for `path`.

File: habitat/tasks/vln/vln.py [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+    r"""Specification of episode that includes initial position and rotation of
+    agent, goal specifications, instruction specifications, and optional shortest paths.
+
+    Args:
+        scene_id: id of scene inside the simulator.
+        start_position: numpy ndarray containing 3 entries for (x, y, z).
+        start_rotation: numpy ndarray with 4 entries for (x, y, z, w)
+            elements of unit quaternion (versor) representing agent 3D
+            orientation.
+        instruction: single instruction guide to goal.
```
Comment (mathfac) [inline habitat/tasks/vln/vln.py]:
```suggestion
        instruction: single natural language instruction guide to goal.
```

File: habitat/tasks/vln/vln.py [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+            "tokens": episode.instruction.instruction_tokens,
+            "trajectory_id": episode.trajectory_id,
+        }
+
+    def get_observation(self, **kwargs):
+        return self._get_observation(**kwargs)
+
+
+@registry.register_task(name="VLN-v0")
+class VLNTask(NavigationTask):
```
Comment (mathfac) [inline habitat/tasks/vln/vln.py]: Extensive doc string should be added with concise description of the task.

File: habitat/datasets/vln/r2r_vln_dataset.py [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+from habitat.core.logging import logger
+from habitat.core.registry import registry
+from habitat.datasets.utils import VocabDict
+from habitat.tasks.nav.nav import NavigationGoal, ShortestPathPoint
+from habitat.tasks.vln.vln import InstructionData, VLNEpisode
+
+CONTENT_SCENES_PATH_FIELD = "content_scenes_path"
+DEFAULT_SCENE_PATH_PREFIX = "data/scene_datasets/mp3d/"
+
+R2R_TRAIN_EPISODES = 10837
```
Comment (mathfac) [inline habitat/datasets/vln/r2r_vln_dataset.py]: This constants are never used.

File: examples/vln_shortest_path_follower_example.py [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+        info, observations["heading"], im.shape[0]
+    )
+    output_im = np.concatenate((im, top_down_map), axis=1)
+    cv2.imwrite("title.jpg", output_im)
+    observations["instruction"]["text"]
+    output_im = append_text_to_image(
+        output_im, observations["instruction"]["text"]
+    )
+    images.append(output_im)
+
```
Comment (mathfac) [inline examples/vln_shortest_path_follower_example.py]: Are all functions above required? Can we import and reuse functions from other examples or util functions?

File: examples/vln_benchmark.py [hunk 23]
```diff
@@ -0,0 +1,106 @@
...
+            observations = env.step(best_action)
+
+    agg_metrics: Dict = defaultdict(float)
+    for m, v in env.get_metrics().items():
+        agg_metrics[m] += v
+    avg_metrics = {k: v / len(env.episodes) for k, v in agg_metrics.items()}
+    return avg_metrics
+
+
+def main():
```
Comment (mathfac) [inline examples/vln_benchmark.py]: Ideally, we would like to have same launcher for different task benchmarks. It should be possible to incorporate task config into `examples/benchmark.py` and make it support VLN as well. But most probably one more registry needed for supported agents/baselines. Let's create follow up issue with merging `benchmark.py` and `vln_benchmark.py`.

File: habitat/tasks/vln/vln.py (hunk 18 above)
Comment (jacobkrantz) [reply habitat/tasks/vln/vln.py]: The path defined in the R2R dataset is a sequence of [x, y, z] positions. The `shortest_paths` in NavigationEpisode uses `ShortestPathPoint` which requires a rotation to be specified as well. To reconcile, we could make the rotation in `ShortestPathPoint` optional or keep a separate `path` as we have now. Any suggestions on handling that?

File: examples/vln_shortest_path_follower_example.py [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+from habitat.utils.visualizations.utils import images_to_video
+
+cv2 = try_cv2_import()
+
+IMAGE_DIR = os.path.join("examples", "images")
+if not os.path.exists(IMAGE_DIR):
+    os.makedirs(IMAGE_DIR)
+
+
+def append_text_to_image(orig_img, text):
```
Comment (mathfac) [inline examples/vln_shortest_path_follower_example.py]: Maybe move to `habitat.utils.visualizations.utils`.

File: habitat/datasets/vln/r2r_vln_dataset.py [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+        with gzip.open(dataset_filename, "rt") as f:
+            self.from_json(f.read(), scenes_dir=config.SCENES_DIR)
+
+    def from_json(
+        self, json_str: str, scenes_dir: Optional[str] = None
+    ) -> None:
+
+        deserialized = json.loads(json_str)
+
+        # Done for the serialization test
```
Comment (mathfac) [inline habitat/datasets/vln/r2r_vln_dataset.py]: What is the reason that we need this if? Why we don't have `deserialized["instruction_vocab"]["word_list"]` in released version of dataset?

File: habitat/tasks/vln/vln.py (hunk 18 above)
Comment (mathfac) [reply habitat/tasks/vln/vln.py]: Make the rotation in `ShortestPathPoint` optional sounds good. You can re-use visualization we have for shortest path already.

Comment (erikwijmans) [reply habitat/tasks/vln/vln.py]: Probably makes more sense to name this reference path (I believe that is what it is called in the VLN dataset).  That'll keep the shortest-path field open for an oracle trajectory for imitation learning.

File: configs/tasks/vln_r2r.yaml (hunk 9 above)
Comment (steflee) [reply configs/tasks/vln_r2r.yaml]: Actually, the VLN task specifies 3m as the success window. Why are we lowering this here?

File: habitat/tasks/vln/vln.py (hunk 18 above)
Comment (mathfac) [reply habitat/tasks/vln/vln.py]: Agreed, `reference_path` then.

File: habitat/tasks/vln/vln.py [hunk 26]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
```
Comment (mathfac) [inline habitat/tasks/vln/vln.py]: You can include your names as owners/maintainers of the VLN task here as well as reflect your contribution in README section.

File: configs/tasks/vln_r2r.yaml [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+  AGENT_0:
+    SENSORS: ['RGB_SENSOR', 'DEPTH_SENSOR']
+  FORWARD_STEP_SIZE: 0.25
+  TURN_ANGLE: 15
+  HABITAT_SIM_V0:
+    GPU_DEVICE_ID: 0
+  RGB_SENSOR:
+    WIDTH: 256
+    HEIGHT: 256
+    HFOV: 45
```
Comment (erikwijmans) [inline configs/tasks/vln_r2r.yaml]: Why 45 degrees?  That seems incredibly narrow

File: configs/datasets/vln/mp3d_r2r.yaml [hunk 28]
```diff
@@ -0,0 +1,5 @@
+DATASET:
+  TYPE: R2RVLN-v1
+  SPLIT: train
+  DATA_PATH: "data/datasets/vln/r2r/v1/{split}/{split}.json.gz"
```
Comment (mathfac) [inline configs/datasets/vln/mp3d_r2r.yaml]: To be consistent with our dataset paths let change path to everywhere:
```suggestion
  DATA_PATH: "data/datasets/vln/mp3d/r2r/v1/{split}/{split}.json.gz"
```

File: habitat/datasets/vln/r2r_vln_dataset.py [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+from habitat.datasets.utils import VocabDict
+from habitat.tasks.nav.nav import NavigationGoal, ShortestPathPoint
+from habitat.tasks.vln.vln import InstructionData, VLNEpisode
+
+DEFAULT_SCENE_PATH_PREFIX = "data/scene_datasets/mp3d/"
+
+R2R_VAL_SEEN_EPISODES = 781
+
+
+@registry.register_dataset(name="R2RVLN-v1")
```
Comment (mathfac) [inline habitat/datasets/vln/r2r_vln_dataset.py]: We should register it `MP3DR2RVLN-v1` or change doc string not attach it to MP3D only. From my experience this class can be scene dataset agnostic.
```suggestion
@registry.register_dataset(name="R2RVLN-v1")
```

Comment (jacobkrantz) [reply habitat/datasets/vln/r2r_vln_dataset.py]: I changed the doc string as the class is not tied to MP3D.

File: habitat/tasks/vln/vln.py (hunk 26 above)
Comment (jacobkrantz) [reply habitat/tasks/vln/vln.py]: I added our names in this file. For the README section, we can wait until we have a paper (as per discussion).

**Comment by jacobkrantz** [comment]
@erikwijmans @mathfac back to you!

**Comment by mathfac** [comment]
@jacobkrantz and @koshyanand thank you for the contribution of such popular Embodied AI task and following all the comments.

**Pull request merged**